src/Juvix/Compiler/Asm/Extra.hs

```diff
@@ -30,11 +30,49 @@ validateCode tab args = void . recurse sig args
 validateFunction :: Member (Error AsmError) r => InfoTable -> FunctionInfo -> Sem r ()
 validateFunction tab fi = validateCode tab (argumentsFromFunctionInfo fi) (fi ^. functionCode)
 
-validateInfoTable :: Member (Error AsmError) r => InfoTable -> Sem r ()
-validateInfoTable tab = mapM_ (validateFunction tab) (HashMap.elems (tab ^. infoFunctions))
+validateInfoTable :: Member (Error AsmError) r => InfoTable -> Sem r InfoTable
+validateInfoTable tab = do
+  mapM_ (validateFunction tab) (HashMap.elems (tab ^. infoFunctions))
+  return tab
 
 validate :: InfoTable -> Maybe AsmError
 validate tab =
   case run $ runError $ validateInfoTable tab of
     Left err -> Just err
     _ -> Nothing
+
+computeFunctionStackUsage :: Member (Error AsmError) r => InfoTable -> FunctionInfo -> Sem r FunctionInfo
+computeFunctionStackUsage tab fi = do
+  ps <- snd <$> recurseS sig initialStackInfo (fi ^. functionCode)
+  let maxValueStack = maximum (map fst ps)
+      maxTempStack = maximum (map snd ps)
+  return
+    fi
+      { _functionMaxValueStackHeight = maxValueStack,
+        _functionMaxTempStackHeight = maxTempStack
+      }
+  where
+    sig :: RecursorSig StackInfo r (Int, Int)
+    sig =
+      RecursorSig
+        { _recursorInfoTable = tab,
+          _recurseInstr = \si _ -> return (si ^. stackInfoValueStackHeight, si ^. stackInfoTempStackHeight),
+          _recurseBranch = \si _ l r ->
+            return
+              ( max (si ^. stackInfoValueStackHeight) (max (maximum (map fst l)) (maximum (map fst r))),
+                max (si ^. stackInfoTempStackHeight) (max (maximum (map snd l)) (maximum (map snd r)))
+              ),
+          _recurseCase = \si _ cs md ->
+            return
+              ( max (si ^. stackInfoValueStackHeight) (max (maximum (map (maximum . map fst) cs)) (maybe 0 (maximum . map fst) md)),
+                max (si ^. stackInfoTempStackHeight) (max (maximum (map (maximum . map snd) cs)) (maybe 0 (maximum . map snd) md))
+              )
+        }
+
+computeStackUsage :: Member (Error AsmError) r => InfoTable -> Sem r InfoTable
+computeStackUsage tab = do
+  lst <- mapM (secondM (computeFunctionStackUsage tab)) (HashMap.toList (tab ^. infoFunctions))
+  return tab {_infoFunctions = HashMap.fromList lst}
```

Comment:
```suggestion
  fns <- mapM (computeFunctionStackUsage tab) (tab ^. infoFunctions)
  return tab {_infoFunctions = fns}
```

`HashMap` is Traversable on its values so we don't need to convert to and from list here.

Reply:
Yes, right, I'll change it.